dash-renderer/src/persistence.js

```diff
@@ -0,0 +1,346 @@
+    }
+
+    setItem(key, value) {
+        const setVal = value === void 0 ? UNDEFINED : JSON.stringify(value);
+        this._storage.setItem(storePrefix + key, setVal);
+    }
+
+    removeItem(key) {
+        this._storage.removeItem(storePrefix + key);
+    }
+
+    clear(keyPrefix) {
+        const fullPrefix = storePrefix + keyPrefix;
+        const keyMatch = keyPrefixMatch(fullPrefix);
+        const keysToRemove = [];
+        // 2-step process, so we don't depend on any particular behavior of
+        // key order while removing some
+        for (let i = 0; i < this._storage.length; i++) {
+            const fullKey = this._storage.key(i);
+            if (keyMatch(fullKey)) {
+                keysToRemove.push(fullKey);
+            }
+        }
+        forEach(this._storage.removeItem, keysToRemove);
+    }
+}
+
+class MemStore {
+    constructor() {
+        this._data = {};
+    }
+
+    hasItem(key) {
+        return key in this._data;
+    }
+
+    getItem(key) {
+        // run this storage through JSON too so we know we get a fresh object
+        // each retrieval
+        return JSON.parse(this._data[key] || null);
+    }
+
+    setItem(key, value) {
+        this._data[key] = JSON.stringify(value);
+    }
+
+    removeItem(key) {
+        delete this._data[key];
+    }
+
+    clear(keyPrefix) {
+        forEach(
+            key => delete this._data[key],
+            filter(keyPrefixMatch(keyPrefix), keys(this._data))
+        );
+    }
+}
+
+const stores = {
+    local: new WebStore(window.localStorage),
+    session: new WebStore(window.sessionStorage),
+    memory: new MemStore(),
+};
+
+const noopTransform = {
+    extract: propValue => propValue,
+    apply: (storedValue, _propValue) => storedValue,
+};
+
+const getTransform = (element, propName) =>
+    (element.persistenceTransforms || {})[propName] || noopTransform;
+
+const getNewValKey = (id, propName) => id + '.' + propName;
+const getOriginalValKey = newValKey => newValKey + '.orig';
+const getPersistIdKey = newValKey => newValKey + '.id';
+
+const getProps = layout => {
+    const {props} = layout;
+    const {id, persistence} = props;
+    if (!id || !persistence) {
+        return {props};
+    }
+
+    const element = Registry.resolve(layout);
+    const persisted_props =
+        props.persisted_props || element.defaultProps.persisted_props;
+    const persistence_type =
+        props.persistence_type || element.defaultProps.persistence_type;
+    if (!persisted_props || !persistence_type) {
+        return {};
+    }
+    return {id, props, element, persistence, persisted_props, persistence_type};
+};
+
+export function recordUiEdit(layout, newProps) {
+    const {
+        id,
+        props,
+        element,
+        persistence,
+        persisted_props,
+        persistence_type,
+    } = getProps(layout);
+    if (!persistence) {
+        return;
+    }
+
+    forEachObjIndexed((newPropVal, propName) => {
```

Comment:
The persisted props will be few vs. the actual props of the components. Maybe we could loop on the `persisted_props` instead and `const newPropVal = newProps[propName]` instead? These actions are user driven, so not in a tight loop, either way, I do not have a strong opinion.

Reply:
Most of the time `newProps` only has one key, regardless of how many are in the component. Table is a bit of an outlier in that the 11 `derived_*` props all arrive at once, but it also seems to call `setProps` twice in many cases - once with just input prop, and again with all the `derived_*` - so I think even there it's roughly equal cost either way: as I have it, looping over 1 then 11 items; flipping it we'd loop over 6+6 items.

That said, `if(newProps[propName])` is probably much faster than `if(persisted_props.indexOf(propName) !== -1)` and most components other than table will probably only have one or two persisted props. So yeah, I'll make the change. 🎉

Reply:
https://github.com/plotly/dash/pull/903/commits/b737a3fc9104134b3dfb57c04ef1ad731b1c10bc